File: paperdb/src/androidTest/java/io/paperdb/CustomBookLocationTest.java
package io.paperdb;

import android.support.test.runner.AndroidJUnit4;

import junit.framework.Assert;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.File;
import java.util.HashSet;

import static android.support.test.InstrumentationRegistry.getTargetContext;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;

@RunWith(AndroidJUnit4.class)
public class CustomBookLocationTest {

    @Before
    public void setUp() {
        Paper.INSTANCE.init(getTargetContext().getFilesDir());
    }

    @Test
    public void readWriteDelete_customLocation_with_sub_dirs() {
        //noinspection ResultOfMethodCallIgnored
        deleteRecursive(new File(getTargetContext().getFilesDir() + "/custom"));

        String customLocation = getTargetContext().getFilesDir() + "/custom/location";
        Book book = Paper.INSTANCE.bookOn(customLocation);

        book.write("city", "Victoria");
        assertEquals("Victoria", book.read("city"));

        // Check sub folders created recursively
        File customSubDir = new File(getTargetContext().getFilesDir() + "/custom");
        assertTrue(customSubDir.exists());
        assertTrue(customSubDir.isDirectory());
        assertEquals(1, customSubDir.listFiles().length);
        assertTrue(customSubDir.listFiles()[0].isDirectory());
        assertEquals("location", customSubDir.listFiles()[0].getName());

        File locationSubDir = new File(getTargetContext().getFilesDir() + "/custom/location");
        assertTrue(locationSubDir.exists());
        assertTrue(locationSubDir.isDirectory());
        assertEquals(1, locationSubDir.listFiles().length);
        assertTrue(locationSubDir.listFiles()[0].isDirectory());
        assertEquals("io.paperdb", locationSubDir.listFiles()[0].getName());

        book.delete("city");
        assertFalse(book.contains("city"));
    }

    @Test
    public void readWriteDelete_customLocation_defaultBook() {
        String customLocation = getTargetContext().getFilesDir() + "/custom_location";
        Book bookOnSdcard = Paper.INSTANCE.bookOn(customLocation);
        Book defaultBook = Paper.INSTANCE.book();

        bookOnSdcard.destroy();
        defaultBook.destroy();

        bookOnSdcard.write("city", "Victoria");
        defaultBook.write("city", "Kyiv");

        assertEquals("Victoria", bookOnSdcard.read("city"));
        assertEquals("Kyiv", defaultBook.read("city"));

        bookOnSdcard.delete("city");

        assertFalse(bookOnSdcard.contains("city"));
        assertEquals("Kyiv", defaultBook.read("city"));
    }

    @Test
    public void readWriteDelete_customLocation_customBook() {
        String customLocation = getTargetContext().getFilesDir() + "/custom/location";
        Book bookOnSdcard = Paper.INSTANCE.bookOn(customLocation, "encyclopedia");
        Book defaultBook = Paper.INSTANCE.book("encyclopedia");

        bookOnSdcard.destroy();
        defaultBook.destroy();

        bookOnSdcard.write("city", "Victoria");
        defaultBook.write("city", "Kyiv");

        assertEquals("Victoria", bookOnSdcard.read("city"));
        assertEquals("Kyiv", defaultBook.read("city"));

        bookOnSdcard.delete("city");

        assertFalse(bookOnSdcard.contains("city"));
        assertEquals("Kyiv", defaultBook.read("city"));
    }

    @Test
    public void useCacheFolderAsCustomLocation() {
        String cachePath = getTargetContext().getCacheDir().toString();
        Book cache = Paper.INSTANCE.bookOn(cachePath);
        cache.destroy();

        cache.write("city", "Kyiv");
        assertEquals("Kyiv", cache.read("city"));

        Assert.assertTrue(cache.getPath().endsWith("/io.paperdb.test/cache/io.paperdb"));
    }

    @Test
    public void getPath() {
        Book defaultBookOnSdCard = Paper.INSTANCE.bookOn("/sdcard");
        Book encyclopediaOnSdCard = Paper.INSTANCE.bookOn("/sdcard", "encyclopedia");

        assertEquals("/sdcard/io.paperdb", defaultBookOnSdCard.getPath());
        assertEquals("/sdcard/io.paperdb/key.pt", defaultBookOnSdCard.getPath("key"));
        assertEquals("/sdcard/encyclopedia", encyclopediaOnSdCard.getPath());
        assertEquals("/sdcard/encyclopedia/key.pt", encyclopediaOnSdCard.getPath("key"));
    }

    @Test
    public void bookInstanceIsTheSameForSameLocationAndBookName() {
        Book defaultBook = Paper.INSTANCE.book();
        Book encyclopedia = Paper.INSTANCE.book("encyclopedia");
        Book defaultBookOnSdCard = Paper.INSTANCE.bookOn("/sdcard");
        Book encyclopediaOnSdCard = Paper.INSTANCE.bookOn("/sdcard", "encyclopedia");

        // Check all instances are unique
        HashSet<Book> instanceSet = new HashSet<>();
        instanceSet.add(defaultBook);
        instanceSet.add(encyclopedia);
        instanceSet.add(defaultBookOnSdCard);
        instanceSet.add(encyclopediaOnSdCard);
        assertEquals(4, instanceSet.size());

        assertSame(defaultBook, Paper.INSTANCE.book());
        assertSame(encyclopedia, Paper.INSTANCE.book("encyclopedia"));
        assertSame(defaultBookOnSdCard, Paper.INSTANCE.bookOn("/sdcard"));
        assertSame(encyclopediaOnSdCard, Paper.INSTANCE.bookOn("/sdcard", "encyclopedia"));
    }

    @Test
    public void locationCanBeWithFileSeparatorAtTheEnd() {
        assertEquals("/sdcard/io.paperdb", Paper.INSTANCE.bookOn("/sdcard").getPath());
        assertEquals("/sdcard/io.paperdb", Paper.INSTANCE.bookOn("/sdcard/").getPath());
        assertEquals("/sdcard/encyclopedia",
                Paper.INSTANCE.bookOn("/sdcard", "encyclopedia").getPath());
        assertEquals("/sdcard/encyclopedia",
                Paper.INSTANCE.bookOn("/sdcard/", "encyclopedia").getPath());
    }

    private void deleteRecursive(File fileOrDirectory) {
        if (fileOrDirectory.isDirectory())
            for (File child : fileOrDirectory.listFiles())
                deleteRecursive(child);

        //noinspection ResultOfMethodCallIgnored
        fileOrDirectory.delete();
    }
}

File: paperdb/src/androidTest/java/io/paperdb/CustomBookLocationTest.kt
package io.paperdb

import android.support.test.runner.AndroidJUnit4

import junit.framework.Assert

import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

import java.io.File
import java.util.HashSet

import android.support.test.InstrumentationRegistry.getTargetContext
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertSame
import org.junit.Assert.assertTrue

@RunWith(AndroidJUnit4::class)
class CustomBookLocationTest {

    @Before
    fun setUp() {
        Paper.init(getTargetContext().filesDir)
    }

    @Test
    fun readWriteDelete_customLocation_with_sub_dirs() = runBlocking<Unit> {

        deleteRecursive(File(getTargetContext().filesDir.toString() + "/custom"))

        val customLocation = getTargetContext().filesDir.toString() + "/custom/location"
        val book = Paper.bookOn(customLocation)

        book.write<Any>("city", "Victoria")
        assertEquals("Victoria", book.read("city"))

        // Check sub folders created recursively
        val customSubDir = File(getTargetContext().filesDir.toString() + "/custom")
        assertTrue(customSubDir.exists())
        assertTrue(customSubDir.isDirectory)
        assertEquals(1, customSubDir.listFiles().size.toLong())
        assertTrue(customSubDir.listFiles()[0].isDirectory)
        assertEquals("location", customSubDir.listFiles()[0].name)

        val locationSubDir = File(getTargetContext().filesDir.toString() + "/custom/location")
        assertTrue(locationSubDir.exists())
        assertTrue(locationSubDir.isDirectory)
        assertEquals(1, locationSubDir.listFiles().size.toLong())
        assertTrue(locationSubDir.listFiles()[0].isDirectory)
        assertEquals("io.paperdb", locationSubDir.listFiles()[0].name)

        book.delete("city")
        assertFalse(book.contains("city"))
    }

    @Test
    fun readWriteDelete_customLocation_defaultBook() = runBlocking<Unit> {
        val customLocation = getTargetContext().filesDir.toString() + "/custom_location"
        val bookOnSdcard = Paper.bookOn(customLocation)
        val defaultBook = Paper.book()

        bookOnSdcard.destroy()
        defaultBook.destroy()

        bookOnSdcard.write<Any>("city", "Victoria")
        defaultBook.write<Any>("city", "Kyiv")

        assertEquals("Victoria", bookOnSdcard.read("city"))
        assertEquals("Kyiv", defaultBook.read("city"))

        bookOnSdcard.delete("city")

        assertFalse(bookOnSdcard.contains("city"))
        assertEquals("Kyiv", defaultBook.read("city"))
    }

    @Test
    fun readWriteDelete_customLocation_customBook() = runBlocking<Unit> {
        val customLocation = getTargetContext().filesDir.toString() + "/custom/location"
        val bookOnSdcard = Paper.bookOn(customLocation, "encyclopedia")
        val defaultBook = Paper.book("encyclopedia")

        bookOnSdcard.destroy()
        defaultBook.destroy()

        bookOnSdcard.write<Any>("city", "Victoria")
        defaultBook.write<Any>("city", "Kyiv")

        assertEquals("Victoria", bookOnSdcard.read("city"))
        assertEquals("Kyiv", defaultBook.read("city"))

        bookOnSdcard.delete("city")

        assertFalse(bookOnSdcard.contains("city"))
        assertEquals("Kyiv", defaultBook.read("city"))
    }

    @Test
    fun useCacheFolderAsCustomLocation() = runBlocking<Unit> {
        val cachePath = getTargetContext().cacheDir.toString()
        val cache = Paper.bookOn(cachePath)
        cache.destroy()

        cache.write<Any>("city", "Kyiv")
        assertEquals("Kyiv", cache.read("city"))

        Assert.assertTrue(cache.path.endsWith("/io.paperdb.test/cache/io.paperdb"))
    }

    @Test
    fun getPath() {
        val defaultBookOnSdCard = Paper.bookOn("/sdcard")
        val encyclopediaOnSdCard = Paper.bookOn("/sdcard", "encyclopedia")

        assertEquals("/sdcard/io.paperdb", defaultBookOnSdCard.path)
        assertEquals("/sdcard/io.paperdb/key.pt", defaultBookOnSdCard.getPath("key"))
        assertEquals("/sdcard/encyclopedia", encyclopediaOnSdCard.path)
        assertEquals("/sdcard/encyclopedia/key.pt", encyclopediaOnSdCard.getPath("key"))
    }

    @Test
    fun bookInstanceIsTheSameForSameLocationAndBookName() {
        val defaultBook = Paper.book()
        val encyclopedia = Paper.book("encyclopedia")
        val defaultBookOnSdCard = Paper.bookOn("/sdcard")
        val encyclopediaOnSdCard = Paper.bookOn("/sdcard", "encyclopedia")

        // Check all instances are unique
        val instanceSet = HashSet<Book>()
        instanceSet.add(defaultBook)
        instanceSet.add(encyclopedia)
        instanceSet.add(defaultBookOnSdCard)
        instanceSet.add(encyclopediaOnSdCard)
        assertEquals(4, instanceSet.size.toLong())

        assertSame(defaultBook, Paper.book())
        assertSame(encyclopedia, Paper.book("encyclopedia"))
        assertSame(defaultBookOnSdCard, Paper.bookOn("/sdcard"))
        assertSame(encyclopediaOnSdCard, Paper.bookOn("/sdcard", "encyclopedia"))
    }

    @Test
    fun locationCanBeWithFileSeparatorAtTheEnd() {
        assertEquals("/sdcard/io.paperdb", Paper.bookOn("/sdcard").path)
        assertEquals("/sdcard/io.paperdb", Paper.bookOn("/sdcard/").path)
        assertEquals("/sdcard/encyclopedia",
            Paper.bookOn("/sdcard", "encyclopedia").path)
        assertEquals("/sdcard/encyclopedia",
            Paper.bookOn("/sdcard/", "encyclopedia").path)
    }

    private fun deleteRecursive(fileOrDirectory: File) {
        if (fileOrDirectory.isDirectory)
            for (child in fileOrDirectory.listFiles())
                deleteRecursive(child)


        fileOrDirectory.delete()
    }
}
